rust/arrow/src/array/array.rs

```diff
@@ -157,7 +157,7 @@ pub trait Array: fmt::Debug + Send + Sync + JsonEqual {
     /// assert_eq!(array.is_null(1), true);
     /// ```
     fn is_null(&self, index: usize) -> bool {
-        self.data().is_null(index)
+        self.data_ref().is_null(index)
```

Comment:
This alone improved the performance a lot, but directly using ArrayData was even better. Should also improve some other kernels.

I think the `data` method is a real performance foot-gun and would be in favor of removing it completely. That makes it obvious to every user whether an `Arc` gets cloned. The cloning itself might not be a problem, but I think it sometimes prohibits other compiler optimizations.

Reply:
Really great finding!

Reply:
FWIW, I agree with you: require an explicit `.clone` from the user (also broadly speaking, not just for `.data()` in particular. We should audit the code base for similar occurrences.

Reply:
I created https://issues.apache.org/jira/browse/ARROW-10683 about removing `data` method.